src/fns.rs

```diff
@@ -0,0 +1,181 @@
+use cloudflare::{
+	endpoints::{
+		dns::{
+			DnsContent, DnsRecord, ListDnsRecords, ListDnsRecordsParams,
+			UpdateDnsRecord, UpdateDnsRecordParams,
+		},
+		zone::{ListZones, ListZonesParams},
+	},
+	framework::{
+		async_api::{ApiClient, Client},
+		auth::Credentials,
+		Environment, HttpApiClientConfig,
+	},
+};
+
+use public_ip::{http, Version};
+
+use std::net::IpAddr;
+
+use anyhow::Result;
+
+use crate::api::Cli;
+use std::sync::Arc;
+
+use crate::clone::Clone_;
+
+use std::collections::HashSet;
+
+pub async fn get_ips() -> Result<(Option<IpAddr>, Option<IpAddr>)> {
+	let (ipv4, ipv6) = tokio::join!(
+		public_ip::addr_with(http::ALL, Version::V4),
+		public_ip::addr_with(public_ip::ALL, Version::V6)
+	);
+
+	if (None, None) == (ipv6, ipv4) {
+		anyhow::bail!("Could not determine your current public IP address.")
+	}
+
+	if let Some(ip) = ipv4 {
+		log::info!("{}", ip);
+	}
+	if let Some(ip) = ipv6 {
+		log::info!("{}", ip);
+	};
+	Ok((ipv4, ipv6))
+}
+
+pub fn get_client(cli: &Cli) -> Result<Client> {
+	let credentials: Credentials = if let Some(token) = cli.token.clone() {
+		Credentials::UserAuthToken { token }
+	} else if let (Some(key), Some(email)) =
+		(cli.key.clone(), cli.email.clone())
+	{
+		log::warn!("API Key & Email combo is deprecated. Please switch to using an API token");
+		Credentials::UserAuthKey { email, key }
+	} else {
+		unreachable!()
+	};
+
+	Client::new(
+		credentials,
+		HttpApiClientConfig::default(),
+		Environment::Production,
+	)
+}
+
+pub async fn get_records(
+	cli: &Cli,
+	client: Arc<Client>,
+) -> Result<Vec<(String, Option<String>, Option<DnsRecord>, Option<DnsRecord>)>>
+{
+	let params = ListZones {
+		params: ListZonesParams::default(),
+	};
+	let zones = client.request(&params);
+	// HACK: make a second call since dns::Zone does not implement Clone upstream
```

Comment:
If I'm not mistaken, I think you can avoid this by cloning the zone_id and only capturing that in the async closure. A quick test shows this builds and seems to still work:

```
diff --git a/src/fns.rs b/src/fns.rs
index 81ff477..35e6469 100644
--- a/src/fns.rs
+++ b/src/fns.rs
@@ -72,19 +72,18 @@ pub async fn get_records(
 	let params = ListZones {
 		params: ListZonesParams::default(),
 	};
-	let zones = client.request(&params);
-	// HACK: make a second call since dns::Zone does not implement Clone upstream
-	let zones2 = client.request(&params).await?.result;
+	let zones = client.request(&params).await?.result;

-	let mut handles = Vec::with_capacity(zones2.len());
-	let mut records = Vec::with_capacity(zones2.len() * 10);
+	let mut handles = Vec::with_capacity(zones.len());
+	let mut records = Vec::with_capacity(zones.len() * 10);

-	for zone in zones.await?.result {
+	for zone in &zones {
 		let client = client.clone();
+		let zone_id = zone.id.to_string();
 		handles.push(tokio::spawn(async move {
 			client
 				.request(&ListDnsRecords {
-					zone_identifier: &zone.id,
+					zone_identifier: zone_id.to_string().as_ref(),
 					params: ListDnsRecordsParams::default(),
 				})
 				.await
@@ -117,10 +116,10 @@ pub async fn get_records(
 		.map(|r| {
 			(
 				r.to_owned(),
-				zones2
-					.iter()
-					.find(|z| r.contains(&z.name))
-					.map(|z| z.id.to_owned()),
+				zones.iter().find(|z| r.contains(&z.name)).map(|z| {
+					log::info!("{}", z.id.to_owned());
+					z.id.to_owned()
+				}),
 				records
 					.iter()
 					.find(|rec| {
```

Reply:
I tried a bunch of ways to try and avoid the second call. At least as I have it the first and second call happen asynchronously so as to avoid the overhead as much as possible.